src/server/pkg/collection/collection.go

```diff
@@ -438,6 +482,38 @@ func (i *iterator) Next(key *string, val proto.Unmarshaler) (ok bool, retErr err
 	return false, nil
 }
 
+// Next() writes a fully qualified key (including the path) to the key value
+func (i *paginatedIterator) Next(key *string, val proto.Unmarshaler) (ok bool, retErr error) {
+	if i.index < len(i.resp.Kvs) {
+		kv := i.resp.Kvs[i.index]
+		i.index++
+
+		*key = string(kv.Key)
+		if err := val.Unmarshal(kv.Value); err != nil {
+			return false, err
+		}
+
+		return true, nil
+	}
+	if len(i.resp.Kvs) == 0 {
+		//empty response
+		return false, nil
+	}
+	// Reached end of resp, try for another page
+	fromKey := string(i.resp.Kvs[len(i.resp.Kvs)-1].Key)
+	resp, err := i.etcdClient.Get(i.ctx, fromKey, etcd.WithRange(i.endKey), etcd.WithLimit(QueryPaginationLimit))
+	if err != nil {
+		return false, err
+	}
+	if len(resp.Kvs) == 1 {
+		// Only got the from key, there are no more kvs to fetch from etcd
+		return false, nil
+	}
```

Comment:
This case seems a bit weird to me. Why does it return a single KV if there are no more Kvs left and shouldn't we do something with that single KV it returned?

Reply:
Yea, so the way the `fromKey` works is that when you specify that in a query, it returns that key as well. So if the previous 'page' ended on the key that is actually the last key, the subsequent query will return 1 result.

Reply:
I see, and this won't trip us up if there really only is `1` total key because we won't even get to this. That call to `Get` is above.